creds/src/lib.rs

```diff
@@ -391,22 +392,67 @@ pub fn create_show_proof(client_state: &mut ClientState<ECPairing>, range_pk : &
+// TODO: with SD and DB, that starts looking a lot like the JWT function; merge!
+pub fn create_show_proof_mdl(client_state: &mut ClientState<ECPairing>, range_pk : &RangeProofPK<ECPairing>, proof_spec: &ProofSpec, io_locations: &IOLocations) -> Result<ShowProof<ECPairing>, Box<dyn Error>>
 {
     // Create Groth16 rerandomized proof for showing
+
+    let proof_spec = create_proof_spec_internal(proof_spec, &client_state.config_str)?;
+
+    // commit the expiration date (for non-expired range proof)
     let valid_until_value_pos = io_locations.get_io_location("valid_until_value").unwrap();
-    let dob_value_pos = io_locations.get_io_location("dob_value").unwrap();
-    
-    let mut io_types = vec![PublicIOType::Revealed; client_state.inputs.len()];
+    let mut io_types = vec![PublicIOType::Hidden; client_state.inputs.len()];
     io_types[valid_until_value_pos - 1] = PublicIOType::Committed;
-    io_types[dob_value_pos - 1] = PublicIOType::Committed;
+    // for each range proofed attribute, set the position to Committed
+    for (attr, _) in &proof_spec.range_over_year {
+        let io_loc = io_locations.get_io_location(&format!("{}_value", &attr)).unwrap();
+        io_types[io_loc - 1] = PublicIOType::Committed;
+    }
+
+    for i in io_locations.get_public_key_indices() {
+        io_types[i] = PublicIOType::Revealed;
+    }
+
+    // For the attributes revealed as field elements, we set the position to Revealed and send the value
+    let mut revealed_inputs = vec![];
+    for attr in &proof_spec.revealed {
+        let io_loc = io_locations.get_io_location(&format!("{}_value", &attr)).unwrap();
+        io_types[io_loc - 1] = PublicIOType::Revealed;
+        revealed_inputs.push(client_state.inputs[io_loc - 1]);
+    }
+
+    // For the attributes revealed as digests, we provide the preimage, the verifier will hash it to get the field element
+    let mut revealed_preimages = serde_json::Map::new();
+    for attr in &proof_spec.hashed {
+        let io_loc = match io_locations.get_io_location(&format!("{}_digest", &attr)) {
+            Ok(loc) => loc,
+            Err(_) => {
+                return_error!(
+                    format!("Asked to reveal hashed attribute {}, but did not find it in io_locations\nIO locations: {:?}", attr, io_locations.get_all_names()));
+            }
+        };        
+
+        io_types[io_loc - 1] = PublicIOType::Revealed;
+
+        if client_state.aux.is_none() {
+            return_error!(format!("Proof spec asked to reveal hashed attribute {}, but client state is missing aux data", attr));
+        }
+        let aux = serde_json::from_str::<Value>(client_state.aux.as_ref().unwrap()).unwrap();
+        let aux = aux.as_object().unwrap();
+        revealed_preimages.insert(attr.clone(), json!(aux[attr].clone().to_string()));
+    }
 
-    let revealed_inputs : Vec<<ECPairing as Pairing>::ScalarField> = vec![];
+    if proof_spec.device_bound {
+        // FIXME: implement
```

Comment:
Could add the unimplemented macro here.
https://doc.rust-lang.org/std/macro.unimplemented.html